README.md
# IPG Data Import Service

## Description
Go service that downloads user and product data from CSV and updates/creates users and products in Zapsi

* Periodocity of download: 10 minutes
* CSV directory: \\server-zapsi\zapsi_ipg_data
   * user file has to have 'zam' in its name
   * product file has to have 'prod' in its name
* service created with ``sc.exe CREATE ipg_data_import_service binpath="c:\Zapsi\ipg_data_import_service
\ipg_data_import_service_windows.exe"``
    * added automatic startup
    * security changed to run under "Zapsi" user
   
### User mapping:

|CSV Name|Zapsi Name, user table|
|------------------|------------------|
|-|FirstName|
|Jméno|Name|
|osobní číslo|Login|
|bckRFID Kód|Rfid changed to decimal (see below)|
|PIN|Barcode|
|PIN|Pin|
|Typ uživatele|UserTypeID|
|-|Email|
|-|Phone|
|1|UserRoleID|
|bckRFID Kód|bckRfid|

### Product mapping:
    
|CSV Name|Zapsi Name, product table|
|------------------|------------------|
|název produktu|Name|
|kód produktu|Barcode|
|čas cyklu|Cycle|
|-|IdleFromTime|
|1|ProductStatusID|
|-|Deleted|
|skupina produktů|ProductGroupID|
|kavita|Cavity|
|čas přípravy|-|
|zmetkovitost|-|


© 2020 Petr Jahoda


CHANGELOG.md
# IPG Data Import Service Changelog

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

Please note, that this project, while following numbering syntax, it DOES NOT
adhere to [Semantic Versioning](http://semver.org/spec/v2.0.0.html) rules.

## Types of changes

* ```Added``` for new features.
* ```Changed``` for changes in existing functionality.
* ```Deprecated``` for soon-to-be removed features.
* ```Removed``` for now removed features.
* ```Fixed``` for any bug fixes.
* ```Security``` in case of vulnerabilities.

## [2021.2.2.17] - 2021-5-17

### Removed
- upx usage from windows binary

### Changed
- directory for csv file on local disk

## [2021.2.2.15] - 2021-5-15

### Changed
- directory for csv file


## [2020.4.2.11] - 2020-11-11

### Added
- when found product group id, prepare time and  scrap percent is updated
- when updating, creating product group, adding cycle from product

## [2020.4.1.26] - 2020-10-26

### Fixed
- fixed leaking goroutine bug when opening sql connections, the right way is this way

## [2020.3.3.19] - 2020-9-19

### Changed
- maps are created with their initial size

## [2020.3.3.18] - 2020-9-18

### Fixed
- when parsing cycle from CSV, proper checking for "9,9" instead of "9.9"

## [2020.3.3.15] - 2020-9-15

### Added
- complete functionality
- windows binary

### Changed
- comparing using map, because it is faster, better with big O(1)
- logging to proper folder when running as windows service

## [2020.3.3.14] - 2020-9-14

### Added
- initial commit


main.go
package main

import (
	"github.com/kardianos/service"
	"time"
)

const version = "2021.2.2.17"
const serviceName = "IPG Data Import Service"
const serviceDescription = "Download users and products from CSV file and imports them into Zapsi database"
const downloadInSeconds = 600
const deleteLogsAfter = 240 * time.Hour

var serviceRunning = false
var processRunning = false
var zapsiConfig = "zapsi_uzivatel:zapsi@tcp(localhost:3306)/zapsi2?charset=utf8&parseTime=True&loc=Local"

type program struct{}

func main() {
	logInfo("MAIN", serviceName+" ["+version+"] starting...")
	logInfo("MAIN", serviceDescription)
	serviceConfig := &service.Config{
		Name:        serviceName,
		DisplayName: serviceName,
		Description: serviceDescription,
	}
	program := &program{}
	s, err := service.New(program, serviceConfig)
	if err != nil {
		logError("MAIN", "Cannot start: "+err.Error())
	}
	err = s.Run()
	if err != nil {
		logError("MAIN", "Cannot start: "+err.Error())
	}
}

func (p *program) Start(service.Service) error {
	logInfo("MAIN", serviceName+" ["+version+"] started")
	serviceRunning = true
	go p.run()
	return nil
}

func (p *program) Stop(service.Service) error {
	serviceRunning = false
	if processRunning {
		logInfo("MAIN", serviceName+" ["+version+"] stopping...")
		time.Sleep(1 * time.Second)
	}
	logInfo("MAIN", serviceName+" ["+version+"] stopped")
	return nil
}

func (p *program) run() {
	logDirectoryFileCheck("MAIN")
	createConfigIfNotExists()
	loadSettingsFromConfigFile()
	for serviceRunning {
		processRunning = true
		start := time.Now()
		logInfo("MAIN", serviceName+" ["+version+"] running")
		importData()
		sleepTime := downloadInSeconds*time.Second - time.Since(start)
		logInfo("MAIN", "Sleeping for "+sleepTime.String())
		time.Sleep(sleepTime)
		deleteOldLogFiles()
		processRunning = false
	}
}


config.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

type Config struct {
	DatabaseType string
	IpAddress    string
	DatabaseName string
	Port         string
	Login        string
	Password     string
}

func createConfigIfNotExists() {
	dir := GetDirectory()
	configDirectory := filepath.Join(dir, "config")
	configFileName := "config.json"
	configFullPath := strings.Join([]string{configDirectory, configFileName}, "/")
	if _, checkPathError := os.Stat(configFullPath); checkPathError == nil {
		logInfo("MAIN", "Config file already exists")
	} else if os.IsNotExist(checkPathError) {
		logInfo("MAIN", "Config file does not exist, creating")
		mkdirError := os.MkdirAll(configDirectory, 0777)
		if mkdirError != nil {
			logError("MAIN", "Unable to create directory for config file: "+mkdirError.Error())
		} else {
			logInfo("MAIN", "Directory for config file created")
			data := Config{
				DatabaseType: "mysql",
				IpAddress:    "localhost",
				DatabaseName: "zapsi2",
				Port:         "3306",
				Login:        "zapsi_uzivatel",
				Password:     "zapsi",
			}
			file, _ := json.MarshalIndent(data, "", "  ")
			writingError := ioutil.WriteFile(configFullPath, file, 0666)
			logInfo("MAIN", "Writing data to JSON file")
			if writingError != nil {
				logError("MAIN", "Unable to write data to config file: "+writingError.Error())
			} else {
				logInfo("MAIN", "Data written to config file")
			}
		}
	} else {
		logError("MAIN", "Config file does not exist")
	}
}

func loadSettingsFromConfigFile() {
	dir := GetDirectory()
	configDirectory := filepath.Join(dir, "config")
	configFileName := "config.json"
	configFullPath := strings.Join([]string{configDirectory, configFileName}, "/")
	ConfigFile := Config{}
	for len(ConfigFile.DatabaseName) == 0 {
		readFile, err := ioutil.ReadFile(configFullPath)
		if err != nil {
			logError("MAIN", "Problem reading config file")
			var err = os.Remove(configFullPath)
			if err != nil {
				logError("MAIN", "Problem deleting file "+configFullPath+", "+err.Error())
				break
			}
			createConfigIfNotExists()
		}
		err = json.Unmarshal(readFile, &ConfigFile)
		if err != nil {
			logError("MAIN", "Problem parsing config file, deleting config file")
			var err = os.Remove(configFullPath)
			if err != nil {
				logError("MAIN", "Problem deleting file "+configFullPath+", "+err.Error())
				break
			}
			createConfigIfNotExists()
		}
	}
	zapsiConfig = ConfigFile.Login + ":" + ConfigFile.Password + "@tcp(" + ConfigFile.IpAddress + ":" + ConfigFile.Port + ")/" + ConfigFile.DatabaseName + "?charset=utf8&parseTime=True&loc=Local"
}


import.go
package main

import (
	"database/sql"
	"encoding/csv"
	"gorm.io/driver/mysql"
	"gorm.io/gorm"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

func importData() {
	timer := time.Now()
	logInfo("MAIN", "Importing process started")
	zapsiUsers, zapsiProducts, downloadedFromZapsi := downloadDataFromZapsi()
	csvUsers, csvProducts, downloadedFromCsvFile := downloadDataFromCsvFile()
	if downloadedFromZapsi && downloadedFromCsvFile {
		logInfo("MAIN", "Zapsi Users: "+strconv.Itoa(len(zapsiUsers)))
		logInfo("MAIN", "Zapsi Products: "+strconv.Itoa(len(zapsiProducts)))
		logInfo("MAIN", "CSV Users: "+strconv.Itoa(len(csvUsers)))
		logInfo("MAIN", "CSV Products: "+strconv.Itoa(len(csvProducts)))
		updatedUsers, createdUsers := processUsers(zapsiUsers, csvUsers)
		updatedProducts, createdProducts := processProducts(zapsiProducts, csvProducts)
		logInfo("MAIN", "Updated users: "+strconv.Itoa(updatedUsers))
		logInfo("MAIN", "Created users: "+strconv.Itoa(createdUsers))
		logInfo("MAIN", "Updated products: "+strconv.Itoa(updatedProducts))
		logInfo("MAIN", "Created products: "+strconv.Itoa(createdProducts))
	}
	logInfo("MAIN", "Importing process complete, time elapsed: "+time.Since(timer).String())
}

func processProducts(zapsiProducts map[string]product, csvProducts []csvProduct) (int, int) {
	timer := time.Now()
	logInfo("MAIN", "Processing products")
	updated := 0
	created := 0
	for _, csvProduct := range csvProducts {
		if serviceRunning {
			_, productInZapsi := zapsiProducts[csvProduct.kodProduktu]
			if productInZapsi {
				updateProductInZapsi(csvProduct)
				updated++
			} else {
				createProductInZapsi(csvProduct)
				created++
			}
		}
	}
	logInfo("MAIN", "Products processed, time elapsed: "+time.Since(timer).String())
	return updated, created
}

func createProductInZapsi(csvProduct csvProduct) {
	logInfo("MAIN", csvProduct.nazevProduktu+": Product does not exist in Zapsi, creating...")
	productGroupId := getProductGroupId(csvProduct)
	db, err := gorm.Open(mysql.Open(zapsiConfig), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return
	}

	cavityAsInt, err := strconv.Atoi(csvProduct.kavita)
	if err != nil {
		cavityAsInt = 0
	}
	cycleAsInt, err := strconv.Atoi(strings.ReplaceAll(csvProduct.casCyklu, ",", "."))
	var cycleAsFloat float64
	if err != nil {
		cycleAsFloat, err = strconv.ParseFloat(strings.ReplaceAll(csvProduct.casCyklu, ",", "."), 64)
		if err != nil {
			cycleAsFloat = 0.0
		}
	} else {
		cycleAsFloat = float64(cycleAsInt)
	}
	var product product
	product.Name = csvProduct.nazevProduktu
	product.Barcode = csvProduct.kodProduktu
	product.Cycle = cycleAsFloat
	product.ProductGroupID = productGroupId
	product.ProductStatusID = 1
	product.Cavity = cavityAsInt
	db.Save(&product)
}

func updateProductInZapsi(csvProduct csvProduct) {
	productGroupId := getProductGroupId(csvProduct)
	db, err := gorm.Open(mysql.Open(zapsiConfig), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return
	}
	cavityAsInt, err := strconv.Atoi(csvProduct.kavita)
	if err != nil {
		cavityAsInt = 0
	}
	cycleAsInt, err := strconv.Atoi(strings.ReplaceAll(csvProduct.casCyklu, ",", "."))
	var cycleAsFloat float64
	if err != nil {
		cycleAsFloat, err = strconv.ParseFloat(strings.ReplaceAll(csvProduct.casCyklu, ",", "."), 64)
		if err != nil {
			cycleAsFloat = 0.0
		}
	} else {
		cycleAsFloat = float64(cycleAsInt)
	}

	db.Model(&product{}).Where(user{Barcode: csvProduct.kodProduktu}).Updates(product{
		Name:           csvProduct.nazevProduktu,
		Cycle:          cycleAsFloat,
		Cavity:         cavityAsInt,
		ProductGroupID: productGroupId,
	})
}

func getProductGroupId(csvProduct csvProduct) int {
	prepareTimeAsFloat, err := strconv.ParseFloat(csvProduct.casPripravy, 64)
	if err != nil {
		logError("MAIN", "Problem parsing cavity for "+csvProduct.nazevProduktu+": "+csvProduct.kavita)
		prepareTimeAsFloat = 0.0
	}
	scrapPercentAsFloat, err := strconv.ParseFloat(csvProduct.zmetkovitost, 64)
	if err != nil {
		logError("MAIN", "Problem parsing cavity for "+csvProduct.nazevProduktu+": "+csvProduct.kavita)
		scrapPercentAsFloat = 0.0
	}
	cycleAsInt, err := strconv.Atoi(strings.ReplaceAll(csvProduct.casCyklu, ",", "."))
	var cycleAsFloat float64
	if err != nil {
		cycleAsFloat, err = strconv.ParseFloat(strings.ReplaceAll(csvProduct.casCyklu, ",", "."), 64)
		if err != nil {
			cycleAsFloat = 0.0
		}
	} else {
		cycleAsFloat = float64(cycleAsInt)
	}
	db, err := gorm.Open(mysql.Open(zapsiConfig), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()

	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return 1
	}
	var existingProductGroup productGroup
	db.Where("Name like ?", csvProduct.skupinaProduktu).Find(&existingProductGroup)
	if existingProductGroup.OID > 0 {
		logInfo("MAIN", "Updating product group "+csvProduct.skupinaProduktu)
		existingProductGroup.PrepareTime = prepareTimeAsFloat
		existingProductGroup.ScrapPercent = scrapPercentAsFloat
		existingProductGroup.Cycle = cycleAsFloat
		db.Save(&existingProductGroup)
		return existingProductGroup.OID
	}
	logInfo("MAIN", "Product group "+csvProduct.skupinaProduktu+" does not exist, creating ...")
	var newProductGroup productGroup
	newProductGroup.Name = csvProduct.skupinaProduktu
	newProductGroup.PrepareTime = prepareTimeAsFloat
	newProductGroup.ScrapPercent = scrapPercentAsFloat
	existingProductGroup.Cycle = cycleAsFloat
	db.Save(&newProductGroup)
	var brandNewProductGroup productGroup
	db.Where("Name like ?", csvProduct.skupinaProduktu).Find(&brandNewProductGroup)
	if brandNewProductGroup.OID > 0 {
		return brandNewProductGroup.OID
	}
	return 1
}

func processUsers(zapsiUsers map[string]user, csvUsers []csvUser) (int, int) {
	timer := time.Now()
	logInfo("MAIN", "Processing users")
	updated := 0
	created := 0
	for _, csvUser := range csvUsers {
		if serviceRunning {
			_, userInZapsi := zapsiUsers[csvUser.osobniCislo]
			if userInZapsi {
				updateUserInZapsi(csvUser, zapsiUsers[csvUser.osobniCislo])
				updated++
			} else {
				createUserInZapsi(csvUser)
				created++
			}
		}
	}
	logInfo("MAIN", "Users processed, time elapsed: "+time.Since(timer).String())
	return updated, created
}

func updateUserInZapsi(csvUser csvUser, zapsiUser user) {
	db, err := gorm.Open(mysql.Open(zapsiConfig), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return
	}
	userTypeIdToInsert := getUserTypeID(csvUser)
	rfidInDecimal := getRfidInHex(csvUser.rfidKod)
	db.Model(&user{}).Where(user{Login: zapsiUser.Login}).Updates(user{
		Name:       csvUser.jmeno,
		Login:      csvUser.osobniCislo,
		Rfid:       rfidInDecimal,
		Pin:        csvUser.pin,
		Barcode:    csvUser.pin,
		UserTypeID: sql.NullInt32{Int32: int32(userTypeIdToInsert), Valid: true},
		BckRfid:    csvUser.rfidKod,
	})
}

func getRfidInHex(rfid string) string {
	trimmedRfid := rfid[0:8]
	originalData := "0123456789ABCDEF"
	replacementData := []rune("084C2A6E195D3B7F")
	rfidConverted := ""
	for _, element := range trimmedRfid {
		index := strings.IndexRune(originalData, element)
		rfidConverted += string(replacementData[index])
	}
	result, err := strconv.ParseUint(rfidConverted, 16, 32)
	if err != nil {
		logError("MAIN", "Problem converting rfid to decimal: "+err.Error())
	}
	resultAsString := strconv.FormatUint(result, 10)

	resultLength := len(resultAsString)
	difference := 10 - resultLength
	for i := 0; i < difference; i++ {
		resultAsString = "0" + resultAsString
	}
	return resultAsString
}

func getUserTypeID(csvUser csvUser) int {
	userTypeIdToInsert := 1
	if strings.Contains(csvUser.typUzivatele, "Mistr") {
		userTypeIdToInsert = 3
	} else if strings.Contains(csvUser.typUzivatele, "Seřizovač") {
		userTypeIdToInsert = 2
	} else if strings.Contains(csvUser.typUzivatele, "Údržbář") {
		userTypeIdToInsert = 4
	} else if strings.Contains(csvUser.typUzivatele, "Supervizor") {
		userTypeIdToInsert = 5
	}
	return userTypeIdToInsert
}

func createUserInZapsi(csvUser csvUser) {
	logInfo("MAIN", csvUser.jmeno+": User does not exist in Zapsi, creating...")
	db, err := gorm.Open(mysql.Open(zapsiConfig), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return
	}
	userTypeIdToInsert := getUserTypeID(csvUser)
	rfidInDecimal := getRfidInHex(csvUser.rfidKod)
	var user user
	user.Login = csvUser.osobniCislo
	user.Name = csvUser.jmeno
	user.Rfid = rfidInDecimal
	user.Barcode = csvUser.pin
	user.Pin = csvUser.pin
	user.UserTypeID = sql.NullInt32{Int32: int32(userTypeIdToInsert), Valid: true}
	user.UserRoleID = sql.NullInt32{Int32: 2, Valid: true}
	user.BckRfid = csvUser.rfidKod
	db.Save(&user)
}

func downloadDataFromCsvFile() ([]csvUser, []csvProduct, bool) {
	timer := time.Now()
	logInfo("MAIN", "Downloading data from Csv")
	var files []string
	//root := "\\\\zapsi.ipg.local\\zapsi_ipg_data"
	root := "C:\\Zapsi\\zapsi ipg data"
	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		files = append(files, path)
		return nil
	})
	if err != nil {
		logError("MAIN", err.Error())
	}
	var csvUsers []csvUser
	var csvProducts []csvProduct
	for _, file := range files {
		if strings.Contains(file, "zam") {
			logInfo("MAIN", "Processing users from "+file)
			csvFile, err := os.Open(file)
			if err != nil {
				logError("MAIN", "Cannot open file: "+err.Error())
			} else {
				r := csv.NewReader(csvFile)
				r.Comma = ';'
				skipFirstLine := true
				for {
					record, err := r.Read()
					if err == io.EOF {
						break
					}
					if err != nil {
						logError("MAIN", "Cannot read file: "+err.Error())
						break
					}
					if skipFirstLine {
						logInfo("MAIN", "Skipping first line containing header")
						skipFirstLine = false
					} else {
						var csvUser csvUser
						if len(record) > 4 {
							if len(record[1]) > 1 {
								csvUser.jmeno = record[0]
								csvUser.osobniCislo = record[1]
								csvUser.rfidKod = record[2]
								csvUser.pin = record[3]
								csvUser.typUzivatele = record[4]
								csvUsers = append(csvUsers, csvUser)
							}
						}
					}
				}
			}

		} else if strings.Contains(file, "prod") {
			logInfo("MAIN", "Processing products from "+file)
			csvFile, err := os.Open(file)
			if err != nil {
				logError("MAIN", "Cannot open file: "+err.Error())
			} else {
				r := csv.NewReader(csvFile)
				r.Comma = ';'
				skipFirstLine := true
				for {
					record, err := r.Read()
					if err == io.EOF {
						break
					}
					if err != nil {
						logError("MAIN", "Cannot read file: "+err.Error())
						break
					}
					if skipFirstLine {
						logInfo("MAIN", "Skipping first line containing header")
						skipFirstLine = false
					} else {
						var csvProduct csvProduct
						if len(record) > 6 {
							csvProduct.nazevProduktu = record[0]
							csvProduct.kodProduktu = record[1]
							csvProduct.skupinaProduktu = record[2]
							csvProduct.casCyklu = record[3]
							csvProduct.kavita = record[4]
							csvProduct.casPripravy = record[5]
							csvProduct.zmetkovitost = record[6]
							csvProducts = append(csvProducts, csvProduct)
						}
					}
				}
			}
		}
	}
	logInfo("MAIN", "Csv data downloaded, time elapsed: "+time.Since(timer).String())
	return csvUsers, csvProducts, true
}

func downloadDataFromZapsi() (map[string]user, map[string]product, bool) {
	timer := time.Now()
	logInfo("MAIN", "Downloading data from Zapsi")
	db, err := gorm.Open(mysql.Open(zapsiConfig), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return nil, nil, false
	}
	var users []user
	var products []product
	db.Find(&users)
	db.Find(&products)
	returnProducts := make(map[string]product, len(products))
	returnUsers := make(map[string]user, len(users))
	for _, product := range products {
		returnProducts[product.Barcode] = product
	}
	for _, user := range users {
		returnUsers[user.Login] = user
	}
	logInfo("MAIN", "Zapsi data downloaded, time elapsed: "+time.Since(timer).String())
	return returnUsers, returnProducts, true
}


database.go
package main

import "database/sql"

type user struct {
	OID        int           `gorm:"primary_key;column:OID"`
	Login      string        `gorm:"column:Login"`
	Password   string        `gorm:"column:Password"`
	Name       string        `gorm:"column:Name"`
	FirstName  string        `gorm:"column:FirstName"`
	Rfid       string        `gorm:"column:Rfid"`
	Barcode    string        `gorm:"column:Barcode"`
	Pin        string        `gorm:"column:Pin"`
	Function   string        `gorm:"column:Function"`
	UserTypeID sql.NullInt32 `gorm:"column:UserTypeID"`
	UserRoleID sql.NullInt32 `gorm:"column:UserRoleID"`
	Email      string        `gorm:"column:Email"`
	Phone      string        `gorm:"column:Phone"`
	BckRfid    string        `gorm:"column:BckRfid"`
}

func (user) TableName() string {
	return "user"
}

type product struct {
	OID             int     `gorm:"primary_key;column:OID"`
	Name            string  `gorm:"column:Name"`
	Barcode         string  `gorm:"column:Barcode"`
	Cycle           float64 `gorm:"column:Cycle"`
	IdleFromTime    int     `gorm:"column:IdleFromTime"`
	ProductStatusID int     `gorm:"column:ProductStatusID"`
	Deleted         int     `gorm:"column:Deleted"`
	ProductGroupID  int     `gorm:"column:ProductGroupID"`
	Cavity          int     `gorm:"column:Cavity"`
}

func (product) TableName() string {
	return "product"
}

type productGroup struct {
	OID          int     `gorm:"primary_key;column:OID"`
	Name         string  `gorm:"column:Name"`
	Cycle        float64 `gorm:"column:Cycle"`
	PrepareTime  float64 `gorm:"column:PrepareTime"`
	ScrapPercent float64 `gorm:"column:ScrapPercent"`
}

func (productGroup) TableName() string {
	return "product_group"
}

type csvUser struct {
	jmeno        string
	osobniCislo  string
	rfidKod      string
	pin          string
	typUzivatele string
}

type csvProduct struct {
	nazevProduktu   string
	kodProduktu     string
	skupinaProduktu string
	casCyklu        string
	kavita          string
	casPripravy     string
	zmetkovitost    string
}


log.go
package main

import (
	"fmt"
	"github.com/TwinProduction/go-color"
	"io/ioutil"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

func logInfo(reference, data string) {
	fmt.Println(color.Ize(color.Green, "["+reference+"] --INF-- "+data))
	appendDataToLog("INF", reference, data)
}

func logError(reference, data string) {
	fmt.Println(color.Ize(color.Red, "["+reference+"] --ERR-- "+data))
	appendDataToLog("ERR", reference, data)
	appendDataToErrLog("ERR", reference, data)
}

func logDirectoryFileCheck(reference string) {
	dateTimeFormat := "2006-01-02 15:04:05.000"
	dir := GetDirectory()
	logDirectory := filepath.Join(dir, "log")
	_, checkPathError := os.Stat(logDirectory)
	logDirectoryExists := checkPathError == nil
	if logDirectoryExists {
		fmt.Println(color.Ize(color.Green,
			"["+reference+"] --INF-- "+"Log directory already exists "))
		return
	}
	fmt.Println(color.Ize(color.Yellow, time.Now().Format(dateTimeFormat)+" ["+reference+"] --WRN-- "+"Log directory does not exist, creating"))
	mkdirError := os.MkdirAll(logDirectory, 0777)
	if mkdirError != nil {
		fmt.Println(color.Ize(color.Red, time.Now().Format(dateTimeFormat)+" ["+reference+"] --ERR--"+"Unable to create directory for log file: "+mkdirError.Error()))
		return
	}
}

func appendDataToLog(logLevel string, reference string, data string) {
	dateTimeFormat := "2006-01-02 15:04:05.000"
	logNameDateTimeFormat := "2006-01-02"
	dir := GetDirectory()
	logDirectory := filepath.Join(dir, "log")
	logFileName := reference + " " + time.Now().Format(logNameDateTimeFormat) + ".log"
	logFullPath := strings.Join([]string{logDirectory, logFileName}, "/")
	f, err := os.OpenFile(logFullPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println(color.Ize(color.Yellow, time.Now().Format(dateTimeFormat)+" ["+reference+"] --WAR-- "+"Log file not present: "+err.Error()))
		return
	}
	defer f.Close()
	logData := time.Now().Format("2006-01-02 15:04:05.000   ") + reference + "   " + logLevel + "   " + data
	if _, err := f.WriteString(logData + "\r\n"); err != nil {
		fmt.Println(color.Ize(color.Red, time.Now().Format(dateTimeFormat)+" ["+reference+"] --ERR-- "+"Cannot write to file: "+err.Error()))
	}
}

func appendDataToErrLog(logLevel string, reference string, data string) {
	dateTimeFormat := "2006-01-02 15:04:05.000"
	logNameDateTimeFormat := "2006-01-02"
	dir := GetDirectory()
	logDirectory := filepath.Join(dir, "log")
	logFileName := reference + " " + time.Now().Format(logNameDateTimeFormat) + ".err"
	logFullPath := strings.Join([]string{logDirectory, logFileName}, "/")
	f, err := os.OpenFile(logFullPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println(color.Ize(color.Red, time.Now().Format(dateTimeFormat)+" ["+reference+"] --WAR-- "+"Log file not present: "+err.Error()))
		return
	}
	defer f.Close()
	logData := time.Now().Format("2006-01-02 15:04:05.000   ") + reference + "   " + logLevel + "   " + data
	if _, err := f.WriteString(logData + "\r\n"); err != nil {
		fmt.Println(color.Ize(color.Red, time.Now().Format(dateTimeFormat)+" ["+reference+"] --ERR-- "+"Cannot write to file: "+err.Error()))
	}
}

func deleteOldLogFiles() {
	logInfo("MAIN", "Deleting old log files")
	timer := time.Now()
	directory, err := ioutil.ReadDir("log")
	if err != nil {
		logError("MAIN", "Problem opening log directory")
		return
	}
	now := time.Now()
	logDirectory := filepath.Join(".", "log")
	for _, file := range directory {
		if fileAge := now.Sub(file.ModTime()); fileAge > deleteLogsAfter {
			logInfo("MAIN", "Deleting old log file "+file.Name()+" with age of "+fileAge.String())
			logFullPath := strings.Join([]string{logDirectory, file.Name()}, "/")
			var err = os.Remove(logFullPath)
			if err != nil {
				logError("MAIN", "Problem deleting file "+file.Name()+", "+err.Error())
			}
		}
	}
	logInfo("MAIN", "Old log files deleted, elapsed: "+time.Since(timer).String())
}

func GetDirectory() string {
	var dir string
	if runtime.GOOS == "windows" {
		executable, err := os.Executable()
		if err != nil {
			panic(err)
		}
		dir = filepath.Dir(executable)
	} else {
		dir, _ = os.Getwd()
	}
	return dir
}
